src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,663 @@
+ * THE SOFTWARE.
+ *
+ */
+package com.fulcrumgenomics.fastq
+
+import java.io.File
+import java.nio.file.Path
+import java.util.NoSuchElementException
+
+import com.fulcrumgenomics.cmdline.{ClpGroups, FgBioTool}
+import com.fulcrumgenomics.util.ReadStructure.SubRead
+import com.fulcrumgenomics.util._
+import com.fulcrumgenomics.util.miseq.{Sample, SampleBarcode, SampleSheet}
+import dagr.commons.CommonsDef.{DirPath, FilePath, PathToFastq, _}
+import dagr.commons.util.LazyLogging
+import dagr.sopt.{arg, clp}
+import htsjdk.samtools.SAMFileHeader.SortOrder
+import htsjdk.samtools._
+import htsjdk.samtools.fastq.FastqReader
+import htsjdk.samtools.util._
+
+import scala.collection.mutable
+
+object DemuxFastqs {
+  import SampleBarcode.SampleBarcodeDelimiter
+
+  val UnmatchedSampleId: String = "unmatched"
+
+  /** Custom columns in the Illumina Experiment Manager Sample Sheet. */
+  object CustomSampleSheetHeaderNames {
+    val R1SampleBarcodeBases = "R1_Barcode_Bases"
+    val R2SampleBarcodeBases = "R2_Barcode_Bases"
+  }
+
+  /** Creates the sample output BAM for the given sample.  [setSampleBarcode] should be called first. */
+  def sampleOutputBam(output: Path, sample: Sample): Path = {
+    val sampleBarcode: String = sample.sampleBarcode.getOrElse(throw new IllegalArgumentException("Sample barcode missing for sample: " + sample.sampleName)).toString
+    output.resolve(IOUtil.makeFileNameSafe(String.format("%s.%s.%s.bam", sample.sampleId, sample.sampleName, sampleBarcode)))
+  }
+
+  /** Gets the sample sheet after adding in any custom sample information tracked by DemuxFastqs. */
+  def sampleSheet(sampleSheet: Path, lane: Option[Int] = None): SampleSheet = {
+    new SampleSheet(
+      SampleSheet(sampleSheet, lane=lane)
+        .map(new CustomSample(_))
+        .toSeq
+    )
+  }
+
+  private val solexaQualityConverter: SolexaQualityConverter = SolexaQualityConverter.getSingleton
+
+  /** A custom sample class to support inline sample barcodes for read one and read two. */
+  private[fastq] class CustomSample(sample: Sample)
+    extends Sample(
+      sampleOrdinal      = sample.sampleOrdinal,
+      sampleId           = sample.sampleId,
+      sampleName         = sample.sampleName,
+      libraryId          = sample.libraryId,
+      project            = sample.project,
+      description        = sample.description,
+      lane               = sample.lane,
+      i7IndexBases       = sample.i7IndexBases,
+      i5IndexBases       = sample.i5IndexBases,
+      sampleBarcode      = sample.sampleBarcode,
+      extendedAttributes = sample.extendedAttributes
+    ) {
+    val r1SampleBarcodeBases: Option[String] = extendedAttribute(CustomSampleSheetHeaderNames.R1SampleBarcodeBases)
+    val r2SampleBarcodeBases: Option[String] = extendedAttribute(CustomSampleSheetHeaderNames.R2SampleBarcodeBases)
+
+    /** Returns the sample barcodes in order of sequencing. */
+    override def sampleBarcodeBases: Seq[Option[String]] = {
```

Comment:
Wouldn't it be nicer to return `Seq[String]` and flatten here?  Or do you need to know which ones are Some and which are None?